restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -84,6 +95,96 @@ static Object translateCanonicalKeyToAlternativeKeyIfNeeded(Object canonicalKey,
     }
   }
 
+  private static Object fillInDataDefault(DataSchema schema, Object dataWithoutDefault)
+  {
+    switch (schema.getType())
+    {
+      case RECORD:
+        return fillInDefaultOnRecord((RecordDataSchema) schema, (DataMap) dataWithoutDefault);
+      case TYPEREF:
+        return fillInDefaultOnTyperef((TyperefDataSchema) schema, dataWithoutDefault);
+      case MAP:
+        return fillInDefaultOnMap((MapDataSchema) schema, (DataMap) dataWithoutDefault);
+      case UNION:
+        return fillInDefaultOnUnion((UnionDataSchema) schema, (DataMap) dataWithoutDefault);
+      case ARRAY:
+        return fillInDefaultOnArray((ArrayDataSchema) schema, (DataList) dataWithoutDefault);
+      default:
+        return dataWithoutDefault;
+    }
+  }
+
+  public static DataMap fillInDefaultOnRecord(RecordDataSchema schema, DataMap dataMap)
+  {
+    DataMap dataWithDefault = new DataMap(DataMapBuilder.getOptimumHashMapCapacityFromSize(dataMap.size()));
+    dataWithDefault.putAll(dataMap);
+    for (RecordDataSchema.Field field : schema.getFields())
+    {
+      if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+      {
+        Object fieldData = dataMap.containsKey(field.getName()) ? dataMap.get(field.getName()) : field.getDefault();
+        dataWithDefault.put(field.getName(), fillInDataDefault(field.getType(), fieldData));
+      }
+    }
+    return dataWithDefault;
+  }
+
+  public static DataMap fillInDefaultOnMap(MapDataSchema schema, DataMap dataMap)
+  {
+    DataSchema valueSchema = schema.getValues();
+    DataMap dataWithDefault = new DataMap(dataMap);
```

Comment:
This will check all the entries again.. so better option is to create an empty dataMap and put items inside the loop without checking
(Similar to how you are handling lists)

Reply:
Recommended method used